File: app/src/main/java/com/sjl/bookmark/ui/adapter/BookmarkAdapter.java
package com.sjl.bookmark.ui.adapter;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.sjl.bookmark.R;
import com.sjl.bookmark.entity.table.Bookmark;
import com.sjl.bookmark.ui.activity.BrowserActivity;
import com.sjl.core.util.log.LogUtils;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * TODO
 *
 * @author Kelly
 * @version 1.0.0
 * @filename BookmarkAdapter.java
 * @time 2018/2/7 15:59
 * @copyright(C) 2018 song
 */
public class BookmarkAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private Context context;
    private List<Bookmark> bookmarkList;
    public static final int TYPE_HEADER = 0;  //代表标题
    public static final int TYPE_ITEM = 1;    //代表项目item
    public static final int TYPE_FOOTER = 2;//加载更多条目
    private static final Pattern REGEX_HTTP = Pattern.compile("//(.*?)/");


    /**
     * 正在加载中
     */
    public static final int LOADING_MORE = 0;
    /**
     * 没有加载更多
     */
    public static final int NO_LOAD_MORE = 1;

    /**
     * 上拉加载更多状态-默认为0
     */
    private int mLoadMoreStatus = 0;

    public BookmarkAdapter(Context context, List<Bookmark> bookmarkList) {
        this.context = context;
        this.bookmarkList = bookmarkList;
    }

    public String getTitle(int currentPosition) {
        Bookmark bookmark = bookmarkList.get(currentPosition);
        return bookmark != null ? bookmark.getTitle() : "";
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, final int viewType) {
        //创建不同的 ViewHolder,根据viewtype来判断
        View view;
        if (viewType == TYPE_HEADER) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_bookmark_title, parent, false);
            return new BookmarkTitleHolder(view);
        } else if (viewType == TYPE_ITEM) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_bookmark, parent, false);
            final BookmarkHolder bookmarkHolder = new BookmarkHolder(view);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int position = bookmarkHolder.getAdapterPosition();
                    Intent intent = new Intent(context, BrowserActivity.class);
                    intent.putExtra(BrowserActivity.WEBVIEW_URL, bookmarkList.get(position).getHref());
                    intent.putExtra(BrowserActivity.WEBVIEW_TITLE, bookmarkList.get(position).getText());
                    context.startActivity(intent);

                }
            });
            return bookmarkHolder;
        } else if (viewType == TYPE_FOOTER) {
            view = LayoutInflater.from(context).inflate(R.layout.item_foot, parent,
                    false);
            return new FootViewHolder(view);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        //这里完成数据的绑定
        if (holder instanceof FootViewHolder) {
            LogUtils.i("正在加载中...");
            FootViewHolder footViewHolder = (FootViewHolder) holder;
            switch (mLoadMoreStatus) {
                case LOADING_MORE://加载更多
                    if (footViewHolder.progressBar.getVisibility() == View.GONE) {
                        footViewHolder.progressBar.setVisibility(View.VISIBLE);
                    }
                    footViewHolder.mTvLoadText.setText(R.string.loading);
                    break;
                case NO_LOAD_MORE://没有数据了
                    footViewHolder.progressBar.setVisibility(View.GONE);
                    footViewHolder.mTvLoadText.setText(R.string.no_load_more);
                    break;
                default:
                    break;
            }
            return;
        }
        Bookmark bookmark = bookmarkList.get(position);
        if (holder instanceof BookmarkTitleHolder) {
            BookmarkTitleHolder titleHolder = (BookmarkTitleHolder) holder;
            titleHolder.tv_bookmark_title.setText(bookmark.getTitle());
        } else if (holder instanceof BookmarkHolder) {
            BookmarkHolder itemHolder = (BookmarkHolder) holder;
            itemHolder.tv_text.setText(bookmark.getText());
            itemHolder.iv_icon.setImageBitmap(base64ToBitmap(bookmark.getIcon()));
            String href = bookmark.getHref();
            if (!TextUtils.isEmpty(href)) {
                itemHolder.tv_domain.setText(getSubUtilSimple(href) + "/...");
            } else {
                itemHolder.tv_domain.setText("--");
            }
        }

    }


    @Override
    public int getItemCount() {
        return (bookmarkList != null && !bookmarkList.isEmpty()) ? bookmarkList.size() + 1 : 1;
    }


    @Override
    public int getItemViewType(int position) {
        if (position + 1 == getItemCount()) {
            return TYPE_FOOTER;
        }
        Bookmark bookmark = bookmarkList.get(position);
        if (bookmark.getType() == 0) {
            return TYPE_HEADER;
        } else {
            return TYPE_ITEM;
        }
    }

    public void setLoadingState(int loadingState) {
        this.mLoadMoreStatus = loadingState;
    }

    public void setData(List<Bookmark> bookmarks) {
        this.bookmarkList = bookmarks;
        notifyDataSetChanged();
    }


    static class BookmarkHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.iv_icon)
        ImageView iv_icon;

        @BindView(R.id.tv_text)
        TextView tv_text;

        @BindView(R.id.tv_domain)
        TextView tv_domain;


        public BookmarkHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    static class BookmarkTitleHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.tv_bookmark_title)
        TextView tv_bookmark_title;


        public BookmarkTitleHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    static class FootViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.progressBar)
        ProgressBar progressBar;
        @BindView(R.id.tv_loading_text)
        TextView mTvLoadText;

        public FootViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }

    /**
     * base64转Bitmap
     *
     * @param base64Data
     * @return
     */
    private Bitmap base64ToBitmap(String base64Data) {
        if (TextUtils.isEmpty(base64Data)) {
            return null;
        }
        byte[] bytes = Base64.decode(base64Data, Base64.DEFAULT);
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
    }

    private String getSubUtilSimple(String str) {
        Matcher m = REGEX_HTTP.matcher(str);
        while (m.find()) {
            return m.group(1);
        }
        return "";
    }
}

File: app/src/main/java/com/sjl/bookmark/ui/adapter/BookmarkAdapter.kt
package com.sjl.bookmark.ui.adapter

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.text.TextUtils
import android.util.Base64
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.sjl.bookmark.R
import com.sjl.bookmark.entity.table.Bookmark
import com.sjl.bookmark.ui.activity.BrowserActivity
import com.sjl.core.util.log.LogUtils
import java.util.regex.Pattern

/**
 * TODO
 *
 * @author Kelly
 * @version 1.0.0
 * @filename BookmarkAdapter.java
 * @time 2018/2/7 15:59
 * @copyright(C) 2018 song
 */
class BookmarkAdapter(private val context: Context, private var bookmarkList: List<Bookmark>?) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    /**
     * 上拉加载更多状态-默认为0
     */
    private var mLoadMoreStatus = 0
    fun getTitle(currentPosition: Int): String {
        val bookmark = bookmarkList!![currentPosition]
        return if (bookmark != null) bookmark.title else ""
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        //创建不同的 ViewHolder,根据viewtype来判断
        val view: View
        if (viewType == TYPE_HEADER) {
            view = LayoutInflater.from(parent.context).inflate(R.layout.item_bookmark_title, parent, false)
            return BookmarkTitleHolder(view)
        } else if (viewType == TYPE_ITEM) {
            view = LayoutInflater.from(parent.context).inflate(R.layout.item_bookmark, parent, false)
            val bookmarkHolder = BookmarkHolder(view)
            view.setOnClickListener {
                val position = bookmarkHolder.adapterPosition
                val intent = Intent(context, BrowserActivity::class.java)
                intent.putExtra(BrowserActivity.WEBVIEW_URL, bookmarkList!![position].href)
                intent.putExtra(BrowserActivity.WEBVIEW_TITLE, bookmarkList!![position].text)
                context.startActivity(intent)
            }
            return bookmarkHolder
        } else if (viewType == TYPE_FOOTER) {
            view = LayoutInflater.from(context).inflate(R.layout.item_foot, parent,
                    false)
            return FootViewHolder(view)
        }
        throw RuntimeException("非法viewType：$viewType")
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        //这里完成数据的绑定
        if (holder is FootViewHolder) {
            LogUtils.i("正在加载中...")
            val footViewHolder = holder
            when (mLoadMoreStatus) {
                LOADING_MORE -> {
                    if (footViewHolder.progressBar!!.visibility == View.GONE) {
                        footViewHolder.progressBar!!.visibility = View.VISIBLE
                    }
                    footViewHolder.mTvLoadText!!.setText(R.string.loading)
                }
                NO_LOAD_MORE -> {
                    footViewHolder.progressBar!!.visibility = View.GONE
                    footViewHolder.mTvLoadText!!.setText(R.string.no_load_more)
                }
                else -> {
                }
            }
            return
        }
        val bookmark = bookmarkList!![position]
        if (holder is BookmarkTitleHolder) {
            holder.tv_bookmark_title!!.text = bookmark.title
        } else if (holder is BookmarkHolder) {
            val itemHolder = holder
            itemHolder.tv_text!!.text = bookmark.text
            itemHolder.iv_icon!!.setImageBitmap(base64ToBitmap(bookmark.icon))
            val href = bookmark.href
            if (!TextUtils.isEmpty(href)) {
                itemHolder.tv_domain!!.text = getSubUtilSimple(href) + "/..."
            } else {
                itemHolder.tv_domain!!.text = "--"
            }
        }
    }

    override fun getItemCount(): Int {
        return if (bookmarkList != null && !bookmarkList!!.isEmpty()) bookmarkList!!.size + 1 else 1
    }

    override fun getItemViewType(position: Int): Int {
        if (position + 1 == itemCount) {
            return TYPE_FOOTER
        }
        val bookmark = bookmarkList!![position]
        return if (bookmark.type == 0) {
            TYPE_HEADER
        } else {
            TYPE_ITEM
        }
    }

    fun setLoadingState(loadingState: Int) {
        mLoadMoreStatus = loadingState
    }

    fun setData(bookmarks: List<Bookmark>?) {
        bookmarkList = bookmarks
        notifyDataSetChanged()
    }

    internal class BookmarkHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @JvmField
        @BindView(R.id.iv_icon)
        var iv_icon: ImageView? = null

        @JvmField
        @BindView(R.id.tv_text)
        var tv_text: TextView? = null

        @JvmField
        @BindView(R.id.tv_domain)
        var tv_domain: TextView? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    internal class BookmarkTitleHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @JvmField
        @BindView(R.id.tv_bookmark_title)
        var tv_bookmark_title: TextView? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    internal class FootViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        @JvmField
        @BindView(R.id.progressBar)
        var progressBar: ProgressBar? = null
        
        @JvmField
        @BindView(R.id.tv_loading_text)
        var mTvLoadText: TextView?= null

        init {
            ButterKnife.bind(this, view)
        }
    }

    /**
     * base64转Bitmap
     *
     * @param base64Data
     * @return
     */
    private fun base64ToBitmap(base64Data: String): Bitmap? {
        if (TextUtils.isEmpty(base64Data)) {
            return null
        }
        val bytes = Base64.decode(base64Data, Base64.DEFAULT)
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }

    private fun getSubUtilSimple(str: String): String {
        val m = REGEX_HTTP.matcher(str)
        while (m.find()) {
            return m.group(1)
        }
        return ""
    }

    companion object {
        const val TYPE_HEADER = 0 //代表标题
        const val TYPE_ITEM = 1 //代表项目item
        const val TYPE_FOOTER = 2 //加载更多条目
        private val REGEX_HTTP = Pattern.compile("//(.*?)/")

        /**
         * 正在加载中
         */
        const val LOADING_MORE = 0

        /**
         * 没有加载更多
         */
        const val NO_LOAD_MORE = 1
    }
}